File: java/Components/Strata.Diagnostic.Core/src/strata/diagnostic/core/repository/DiagnosticRepository.java
//////////////////////////////////////////////////////////////////////////////
// DiagnosticRepository.java
//////////////////////////////////////////////////////////////////////////////

package strata.diagnostic.core.repository;

import strata.diagnostic.core.common.DiagnosticSuite;
import strata.diagnostic.core.common.IDiagnostic;

import java.util.Collection;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public
class DiagnosticRepository
    implements IDiagnosticRepository
{
    private final Map<String,IDiagnostic> itsDiagnostics;

    public
    DiagnosticRepository()
    {
        itsDiagnostics = new ConcurrentHashMap<>();
    }

    @Override
    public IDiagnosticRepository
    insert(IDiagnostic diagnostic)
    {
        itsDiagnostics.put(diagnostic.getName(),diagnostic);

        if (diagnostic instanceof DiagnosticSuite)
        {
            DiagnosticSuite suite = (DiagnosticSuite)diagnostic;

            suite
                .getDiagnostics()
                .stream()
                .forEach(d -> insert(d));
        }

        return this;
    }

    @Override
    public IDiagnosticRepository
    remove(IDiagnostic diagnostic)
    {
        itsDiagnostics.remove(diagnostic.getName(),diagnostic);
        return this;
    }

    @Override
    public IDiagnosticRepository
    clear()
    {
        itsDiagnostics.clear();
        return this;
    }

    @Override
    public Optional<IDiagnostic>
    getUnique(String diagnosticName)
    {
        return
            hasUnique(diagnosticName)
                ? Optional.ofNullable(itsDiagnostics.get(diagnosticName))
                : Optional.empty();
    }

    @Override
    public Collection<IDiagnostic>
    getMatching(Predicate<IDiagnostic> predicate)
    {
        return
            itsDiagnostics
                .values()
                .stream()
                .filter(predicate)
                .collect(Collectors.toSet());
    }

    @Override
    public Collection<IDiagnostic>
    getAll()
    {
        return itsDiagnostics.values();
    }

    @Override
    public boolean
    hasUnique(String diagnosticName)
    {
        return itsDiagnostics.containsKey(diagnosticName);
    }

    @Override
    public boolean
    hasMatching(Predicate<IDiagnostic> predicate)
    {
        return
            itsDiagnostics
                .values()
                .stream()
                .anyMatch(predicate);
    }
}

//////////////////////////////////////////////////////////////////////////////

File: csharp/Components/Strata.Diagnostic.Core/Repository/DiagnosticRepository.cs
//  ##########################################################################
//  # File Name: DiagnosticRepository.cs
//  ##########################################################################

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using Strata.Diagnostic.Core.Common;

namespace Strata.Diagnostic.Core.Repository
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// </summary>
    ///  
    public
    class DiagnosticRepository:
        IDiagnosticRepository
    {
        private readonly IDictionary<string,IDiagnostic> itsDiagnostics;

        public 
        DiagnosticRepository()
        {
            itsDiagnostics = new ConcurrentDictionary<string,IDiagnostic>();
        }

        public IDiagnosticRepository 
        Insert(IDiagnostic diagnostic)
        {
            itsDiagnostics.Add(diagnostic.Name,diagnostic);

            if (diagnostic is DiagnosticSuite)
            {
                DiagnosticSuite suite = (DiagnosticSuite)diagnostic;

                foreach (IDiagnostic d in suite.GetDiagnostics())
                    Insert(d);
            }

            return this;
        }

        public IDiagnosticRepository 
        Remove(IDiagnostic diagnostic)
        {
            if (HasUnique(diagnostic.Name))
                itsDiagnostics.Remove(diagnostic.Name);

            return this;
        }

        public IDiagnosticRepository 
        Clear()
        {
            itsDiagnostics.Clear();
            return this;
        }

        public IDiagnostic 
        GetUnique(string diagnosticName)
        {
            return
                HasUnique(diagnosticName)
                    ? itsDiagnostics[diagnosticName]
                    : null;
        }

        public ICollection<IDiagnostic> 
        GetMatching(Func<IDiagnostic,bool> predicate)
        {
            return
                itsDiagnostics
                    .Values
                    .Where(predicate)
                    .ToList();
        }

        public ICollection<IDiagnostic> 
        GetAll()
        {
            return itsDiagnostics.Values;
        }

        public bool 
        HasUnique(string diagnosticName)
        {
            return itsDiagnostics.ContainsKey(diagnosticName);
        }

        public bool
        HasMatching(Func<IDiagnostic,bool> predicate)
        {
            return
                itsDiagnostics
                    .Values
                    .Any(predicate);
        }
    }
}

//  ##########################################################################
